tools/upscaling_coefficients/upscaler_demo.py

```diff
@@ -0,0 +1,809 @@
+    [-0.0145609336341836, -0.0391954528163319, -0.019741250657301437,
+    -0.03620398561701563, -0.01930821727497102], [-0.028882977402423963,
+    0.08045180751196825, 0.5659339775075575, 0.14229549707292577,
+    -0.038669884759381434], [-0.0268506592492741, 0.06284440084948138,
+    0.47416733354946305, 0.11572472115297622, -0.03583254566206615],
+    [-0.012244853215160445, -0.03979621423581153, -0.04046620032608077,
+    -0.03954147855053079, -0.016263925397518374]], [[-0.0010778269600400273,
+    -0.00730937211152405, -0.01288326617180421, -0.00957114638494601,
+    -0.002791220988040242], [-0.010740155274818482, -0.04044013741654317,
+    -0.020308275361446707, -0.031939432194582666, -0.024336140047717],
+    [-0.020702111700092594, 0.03369473939392682, 0.5398093391410693,
+    0.21238920010551754, -0.04562658821385713], [-0.019367833372750342,
+    0.022877571178541396, 0.452206427023829, 0.17567618137781177,
+    -0.04246845133905895], [-0.009019734953997752, -0.039000354140279844,
+    -0.040021010867420236, -0.03838632575427138, -0.02066407496750483]],
+    [[-0.0007159165805851705, -0.006389875713059274, -0.012319067611648212,
+    -0.010202069931803576, -0.0036010963945262565], [-0.007755474877032563,
+    -0.03942402104403912, -0.02118180471068666, -0.0278857406191104,
+    -0.029204772772557758], [-0.015459264122748742, -7.890574314417695e-05,
+    0.4909386897413151, 0.2868949551896562, -0.04562755195174026],
+    [-0.01450239495172347, -0.0057304550220541465, 0.4111930051936302,
+    0.23903219477825297, -0.04314593981787025], [-0.0064664894229144,
+    -0.03663285147762567, -0.03901507994141055, -0.037526010691823306,
+    -0.025194061504750526]], [[-0.0004507227386051219, -0.005422291349995998,
+    -0.01155272937910007, -0.01081748428801323, -0.00447631814814665],
+    [-0.005422291349995998, -0.03680917952171095, -0.022489385288542607,
+    -0.024475522375569655, -0.03324558280295302], [-0.011552729379100074,
+    -0.02248938528854261, 0.4272005024152728, 0.3575770812164143,
+    -0.0378381168526885], [-0.010817484288013228, -0.02447552237556967,
+    0.3575770812164143, 0.29895328454745285, -0.03706352644207268],
+    [-0.00447631814814665, -0.03324558280295302, -0.0378381168526885,
+    -0.03706352644207269, -0.029286133281073243]]]]
+}
+
+
+def convolution(pixels, kernel):
+  """
+  Returns the convolution of `pixels` with `kernel`.
+
+  Uses padding such that the shape of the returned convoluted array is the
+  same as the shape of `pixels`.
+
+  Args:
+    pixels: A [heigth, width]- or [height, width, num_channels]-array
+    representing an image.
+
+    kernel: A [upscaling_factor, upscaling_factor, kernel_size,
+     kernel_size]-array used for the convolution.
+
+  Returns:
+    A [heigth, width]- or [height, width, num_channels]-array representing the
+    convoluted image.
+  """
+  kernel_size = kernel.shape[2]
+  pad_width = kernel_size//2
+  is_multi_channel = len(pixels.shape) > 2
```

Comment:
Would it perhaps make sense to simply normalize the input as follows (such tweaks are OKish-to-do right as the first thing that happens to an input arg in the body)...:

pixels = pixels.reshape(pixels.shape[:2] + (1,))

This eliminates special-casing. From there on, we are guaranteed that this has three indices, and can use more uniform logic for processing.

Note:

>>> z = numpy.zeros([40, 20])
>>> z1 = numpy.zeros([40, 20])
>>> z2 = numpy.zeros([40, 20, 3])
>>> z1r = z1.reshape(z1.shape[:2] + (-1,))
>>> z2r = z2.reshape(z2.shape[:2] + (-1,))
>>> z1r.shape
(40, 20, 1)
>>> z2r.shape
(40, 20, 3)

Reply:
This could be done, but then we would need to do another reshape in the end, because the intention is that the output has the same shape as the input.

Do you still think that this would be a good idea considering this?

Reply:
I would say that a final

return output.reshape(input.shape)

is also rather uniform and nicely conveys the idea that "the output should be the same shape as the input". In that case, the structure would roughly be...

def f(input, ...):
  shaped_input = input.reshape(...)
  ...
  return output.reshape(input)

Reply:
modified the code accordingly; also clarifiedthat the output shape is actually the upscaled shape in the first two dimensions.